### ISSUE TITLE
bug: shutting down in FastAPI with ctrl+c does not hard exit

### ISSUE THREAD

**Issue by elmdecoste** [langfuse/langfuse#4018]
### Describe the bug

When running an an async route hander in FastAPI with pending traces that need to be sent to langfuse, the forced shutdown via ctrl+c does not hard exit the task manager process within the SDK.

### To reproduce

We are sending parallel traces up to Langfuse via an `asyncio.gather()` call that runs through a bunch of different texts at once, and on error we try to shutdown the process when running locally, but the process hangs because the Langfuse task manager (possibly the prompt cache?) is holding a lock.

### SDK and container versions

PyProject version: langfuse = "^2.53.1"
Cloud hosted Langfuse

### Are you interested to contribute a fix for this bug?

No

**Comment by marcklingen** [comment]
Do you instantiate a single Langfuse sdk client? Can you share more details on your implementation?

**Comment by elmdecoste** [comment]
Our implementation just adds the Langfuse callbacks when we use `abatch` or `ainvoke` inside of our langchain wrapper, we never explicitly create a Langfuse sdk client

**Comment by elmdecoste** [comment]
```Exception ignored in atexit callback: <bound method PromptCacheTaskManager.shutdown of <langfuse.prompt_cache.PromptCacheTaskManager object at 0x36096f750>>
Traceback (most recent call last):
  File "/Users/elmd/miniconda3/envs/ndai-backend-poetry/lib/python3.11/site-packages/langfuse/prompt_cache.py", line 122, in shutdown
    consumer.join()
```

This is what we see blocking the shutdown request

**Comment by hassiebp** [comment]
@elmdecoste Are you instantiating a `CallbackHandler` instance on each request? If so, a full Langfuse client is created with its own background threads to deliver events to LF servers. Please refactor your code such that the handler is instantiated only once.

Side note, we have a backlog task to have move the underlying task manager to a singleton that is shared by all Langfuse clients.

**Comment by elmdecoste** [comment]
Yep, each instance gets a callback handler since we instantiate it in a base chain class, we can look into trying to abstract it for now into some singleton although it's not an ideal solution

**Comment by makgah** [comment]
> Hey @elmdecoste! I'm here to help you with any bugs, questions, or contributions you have in mind. Let's squash those bugs together!
> 
> The issue you're experiencing with the Langfuse SDK not exiting immediately when using Ctrl+C in a FastAPI application is likely due to the SDK's task manager ensuring that all pending traces are sent before the process exits. This behavior is by design to prevent data loss.
> 
> To address this, you can use the `langfuse.flush()` method, which is designed to be called at shutdown to ensure all events are sent and all threads are terminated gracefully. This method is blocking and will wait for all events to be flushed before allowing the application to terminate. Here's how you can integrate it:
> 
> ```python
> from contextlib import asynccontextmanager
> from fastapi import FastAPI
> 
> @asynccontextmanager
> async def lifespan(app: FastAPI):
>     yield  # wait until shutdown
> 
>     # Flush all events to be sent to Langfuse on shutdown and terminate all Threads gracefully. This operation is blocking.
>     langfuse.flush()
> 
> app = FastAPI(lifespan=lifespan)
> ```
> 
> This setup should help ensure that your application exits cleanly even when using Ctrl+C, by flushing all pending events before shutdown [[1]](https://langfuse.com/docs/sdk/python/low-level-sdk)[[2]](https://langfuse.com/guides/cookbook/python_sdk_low_level).
> 
> _To continue talking to [Dosu](https://dosu.dev), mention @dosu._

I've experienced the same thing. 
I came to this solution a while back and unfortunately does not work. 
I do not use callbackhandlers and use only simple decorators throughout code. 
The only thing I do with langfuse client is get prompt and update the current observation.

Please advise.

**Comment by hassiebp** [comment]
@makgah Unfortunately I cannot reproduce your issue. Are you on the latest version of our Python SDK? If so, could you please provide a minimum repro example for when you see shutdown hanging?

**Comment by hassiebp** [comment]
This should be resolved on the latest SDK version. Please upgrade and ping us if you experience this issue again 🙏🏾
